### PR TITLE
Groups pages table

### PR DESCRIPTION
**Pull request by liyiy**
## Changes

*Please describe.*  

Adding support for groups (like persons table)

https://user-images.githubusercontent.com/25164963/141602976-28c0b2b5-c250-46cb-9204-18e5b5080896.mov

*If this affects the frontend, include screenshots.*  

## How did you test this code?

*Please describe.*

### CODE REVIEW

File: ee/clickhouse/views/groups.py [hunk 1]
```diff
@@ -18,11 +18,49 @@ class Meta:
         fields = ["group_type", "group_type_index"]
 
 
-class ClickhouseGroupsView(StructuredViewSetMixin, ListModelMixin, viewsets.GenericViewSet):
+class ClickhouseGroupsTypesView(StructuredViewSetMixin, ListModelMixin, viewsets.GenericViewSet):
```
Comment (liyiy) [inline ee/clickhouse/views/groups.py:21]: is it okay to leave this types view in this file since it's so small?

**Comment by liyiy** [comment]
So my localhost broke while testing this so I will update tomorrow morning with screenshots/video, but so far we have the basic table implementation. I still have to add the properties expanded row but other than that this should be good for review

**Comment by mariusandra** [comment]
I'm just your flyby reminder that every logic change, including new logics, [should come with tests](https://posthog.com/docs/contribute/coding-conventions#frontend).

File: ee/clickhouse/views/groups.py (hunk 1 above)
Comment (macobo) [reply ee/clickhouse/views/groups.py:21]: Yup, they are closely related

File: ee/clickhouse/views/groups.py [hunk 2]
```diff
@@ -18,11 +18,49 @@ class Meta:
...
     pagination_class = None
 
+
+class ClickhouseGroupsView(StructuredViewSetMixin, ListModelMixin, viewsets.GenericViewSet):
+    serializer_class = ClickhouseGroupSerializer
+    queryset = None
+    pagination_class = None
+
+    def list(self, request, *args, **kwargs):
+        is_csv_request = self.request.accepted_renderer.format == "csv"
```
Comment (macobo) [inline ee/clickhouse/views/groups.py:33]: Drop the is_csv behavior for now - we can add it later as needed :)

File: posthog/api/utils.py [hunk 3]
```diff
@@ -23,19 +23,19 @@ def get_target_entity(request: request.Request) -> Entity:
         raise ValueError("An entity must be provided for target entity to be determined")
 
 
-def format_next_url(request: request.Request, offset: int, page_size: int):
+def format_next_url(request: request.Request, offset: int, page_size: int, previous: Optional[bool] = None):
```
Comment (macobo) [inline posthog/api/utils.py:26]: `Optional[bool]` doesn't really make sense - you want `previous = False` here instead.

Also the method name is now confusing. Is there a better way to name this function given it handles previous page as well?

File: frontend/src/scenes/groups/Groups.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                        }}
+                    >
+                        <LeftOutlined /> Previous
+                    </Button>
+                    <Button
+                        type="link"
+                        disabled={!groups.next_url}
+                        onClick={() => {
+                            loadGroups(groups.next_url)
+                            window.scrollTo(0, 0)
```
Comment (macobo) [inline frontend/src/scenes/groups/Groups.tsx:63]: Q: Why do we need this?

**Review by macobo — COMMENTED** [review]
Please add images/gifs/explanations to your flow into the PR itself - this makes it easier to understand the flow on a high level.

File: posthog/api/utils.py (hunk 3 above)
Comment (liyiy) [reply posthog/api/utils.py:26]: `format_paginate_url`?

File: frontend/src/scenes/groups/Groups.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            <GroupsTabs />
+            <ResizableTable
+                size="small"
+                columns={columns}
+                rowKey="group_key"
+                loading={groupsLoading}
+                dataSource={groups.results}
+            />
+            {(groups.previous_url || groups.next_url) && (
+                <div style={{ margin: '3rem auto 10rem', width: 200, display: 'flex', alignItems: 'center' }}>
```
Comment (macobo) [inline frontend/src/scenes/groups/Groups.tsx:47]: Let's write a component-specific SCSS file instead and assign this a class.

Given this is really similar with persons, perhaps we can share styles?

File: frontend/src/scenes/groups/groupsListLogic.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    previous_url: string | null
+    results: Group[]
+}
+
+export const groupsListLogic = kea<groupsListLogicType<GroupsPaginatedResponse>>({
+    path: ['groups', 'groupsListLogic'],
+    connect: { values: [teamLogic, ['currentTeamId'], groupsModel, ['groupsEnabled', 'groupTypes']] },
+    actions: () => ({
+        loadGroups: (url?: string | null) => ({ url }),
+        setTab: (tab: string) => ({ tab }),
```
Comment (macobo) [inline frontend/src/scenes/groups/groupsListLogic.ts:20]: Why is tab a string here and a number below? This causes you headaches with typecasting.

File: frontend/src/scenes/persons/Persons.tsx [hunk 7]
```diff
@@ -36,7 +39,8 @@ export function Persons({ cohort }: PersonsProps = {}): JSX.Element {
 
     return (
         <div className="persons-list">
-            {!cohort && <PageHeader title="Persons" />}
+            {!cohort && !groupsEnabled && <PageHeader title="Persons" />}
```
Comment (macobo) [inline frontend/src/scenes/persons/Persons.tsx:42]: Use a ternary here instead:

```
!cohort && (groupsEnabled ? <PageHeader /> : <GroupsTabs />
```

File: frontend/src/scenes/groups/groupsListLogic.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    }),
+    loaders: ({ values }) => ({
+        groups: [
+            { next_url: null, previous_url: null, results: [] } as GroupsPaginatedResponse,
+            {
+                loadGroups: async ({ url }) => {
+                    if (values.groupsEnabled) {
+                        if (!url) {
+                            return await api.get(
+                                `api/projects/${values.currentTeamId}/groups/?group_type_index=${values.currentGroup}`
```
Comment (macobo) [inline frontend/src/scenes/groups/groupsListLogic.ts:30]: You're repeating yourself here.

```
url = url || `api/projects/${values.currentTeamId}/groups/?group_type_index=${values.currentGroup}`
```

File: frontend/src/scenes/groups/groupsListLogic.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+                            )
+                        }
+                        return await api.get(url)
+                    }
+                    return []
```
Comment (macobo) [inline frontend/src/scenes/groups/groupsListLogic.ts:35]: This does not match the structure of `GroupsPaginatedResponse`.

**Review by macobo — COMMENTED** [review]
Copy-pasta, ignore.

File: posthog/api/utils.py (hunk 3 above)
Comment (macobo) [reply posthog/api/utils.py:26]: `format_paginated_url` sounds even better

File: ee/clickhouse/views/groups.py [hunk 10]
```diff
@@ -18,11 +18,48 @@ class Meta:
...
+                SELECT group_type_index, group_key, created_at, group_properties FROM groups
+                INNER JOIN (
+                    SELECT
+                        group_key,
+                        max(created_at) as created_at
+                    FROM groups
+                    WHERE team_id = %(team_id)s AND group_type_index = %(group_type_index)s
+                    GROUP BY group_key
+                    LIMIT 100 OFFSET %(offset)s
+                ) latest_groups
```
Comment (macobo) [inline ee/clickhouse/views/groups.py:47]: The query approach here is wrong. As I mentioned in the latest pairing session, the approach is `group by` and 
 `argMax` instead.

**Comment by macobo** [comment]
Merging this in. Fixed couple of pretty large problems:
1. The query was incorrect and didn't work
2. Pagination did not work 
3. Double pagination was being shown on the frontend (one provided by the Table element)
4. No backend tests

Suggested topic for next pairing session: Creating great PR descriptions and making PRs easy to review.

**Pull request merged**